components/script/dom/readablestreamdefaultreader.rs

```diff
@@ -438,3 +475,119 @@ impl ReadableStreamGenericReader for ReadableStreamDefaultReader {
         Some(self)
     }
 }
+
+type ReadAllBytesSuccessSteps = Rc<dyn Fn(Rc<Promise>, &Vec<u8>)>;
+type ReadAllBytesFailureSteps = Rc<dyn Fn(Rc<Promise>, SafeJSContext, SafeHandleValue)>;
+
+impl js::gc::Rootable for ReadAllBytesFulFillmentHandler {}
+
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+#[cfg_attr(crown, crown::unrooted_must_root_lint::must_root)]
+struct ReadAllBytesFulFillmentHandler {
+    global: DomRoot<GlobalScope>,
+    #[ignore_malloc_size_of = "Rc are hard"]
+    promise: Rc<Promise>,
+    #[ignore_malloc_size_of = "Rc are hard"]
+    #[no_trace]
+    success_steps: ReadAllBytesSuccessSteps,
+    #[ignore_malloc_size_of = "Rc are hard"]
+    #[no_trace]
+    failure_steps: ReadAllBytesFailureSteps,
+    stream: Option<DomRoot<ReadableStream>>,
+    bytes: DomRefCell<Vec<u8>>,
+}
+
+impl Callback for ReadAllBytesFulFillmentHandler {
+    #[cfg_attr(crown, allow(crown::unrooted_must_root))]
+    fn callback(&self, cx: SafeJSContext, v: SafeHandleValue, _realm: InRealm, can_gc: CanGc) {
+        let stream = self
+            .stream
+            .as_ref()
+            .expect("ReadAllBytesFulFillmentHandler has no stream in callback.");
+
+        let is_done = match get_read_promise_done(cx, &v) {
+            Ok(is_done) => is_done,
+            Err(err) => {
+                stream.stop_reading();
+                let reject_handler = Box::new(ReadAllBytesFulFillmentRejectionHandler {
+                    promise: self.promise.clone(),
+                    failure_steps: self.failure_steps.clone(),
+                });
+                reject_handler.failure_error(err, &self.global);
+                return;
+            },
+        };
+
+        if is_done {
+            (self.success_steps)(self.promise.clone(), &self.bytes.borrow());
+        } else {
+            let chunk = match get_read_promise_bytes(cx, &v) {
+                Ok(chunk) => chunk,
+                Err(err) => {
+                    stream.stop_reading();
+                    let reject_handler = Box::new(ReadAllBytesFulFillmentRejectionHandler {
+                        promise: self.promise.clone(),
+                        failure_steps: self.failure_steps.clone(),
+                    });
+                    reject_handler.failure_error(err, &self.global);
+                    return;
+                },
+            };
+
+            let mut bytes = self.bytes.borrow_mut();
+            bytes.extend_from_slice(&chunk);
+
+            let global = stream.global();
```

Comment:
/// [Read-loop](https://streams.spec.whatwg.org/#read-loop) given reader, bytes, successSteps, and failureSteps.